### PR TITLE
Record projected column handles in HiveTableHandle

### PR DESCRIPTION
**Pull request by raunaqmorarka**

### CODE REVIEW

**Comment by findepi** [comment]
cc @losipiuk 

@raunaqmorarka see failures

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ApplyTableScanRedirection.java [hunk 1]
```diff
@@ -92,7 +93,10 @@ public Result apply(TableScanNode scanNode, Captures captures, Context context)
         if (destinationTableHandle.isEmpty()) {
             throw new TrinoException(TABLE_NOT_FOUND, format("Destination table %s from table scan redirection not found", destinationTable));
         }
-        if (destinationTableHandle.get().equals(scanNode.getTable())) {
+
+        TableMetadata tableMetadata = metadata.getTableMetadata(context.getSession(), scanNode.getTable());
+        CatalogSchemaTableName sourceTableName = new CatalogSchemaTableName(tableMetadata.getCatalogName().getCatalogName(), tableMetadata.getTable());
+        if (destinationTable.equals(sourceTableName)) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ApplyTableScanRedirection.java:99]: Add a comment why this change is needed. Could we have a test?

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ApplyTableScanRedirection.java [hunk 2]
```diff
@@ -121,7 +121,7 @@ public Result apply(TableScanNode scanNode, Captures captures, Context context)
                                 destinationTable,
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ApplyTableScanRedirection.java:121]: first 2 commits can be separate PR, thus they would land faster.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ApplyTableScanRedirection.java (hunk 1 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ApplyTableScanRedirection.java:99]: This becomes necessary because the table handle from scan node could contain projected column handles but the destination table handle won't have them. So the redirection could have already taken place but we would fail to detect it due to the difference in table handle.
One of the existing redirection tests failed after I updated MockConnectorTableHandle#equals method

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ApplyTableScanRedirection.java (hunk 2 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ApplyTableScanRedirection.java:121]: Created PR https://github.com/trinodb/trino/pull/7365
Note that the 1st commit is a pre-requisite for the 3rd commit here.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 3]
```diff
@@ -2287,13 +2287,28 @@ public void validateScan(ConnectorSession session, ConnectorTableHandle tableHan
...
 
-        // No pushdown required if all references are simple variables
+        HiveTableHandle hiveTableHandle = (HiveTableHandle) handle;
+        // all references are simple variables
         if (columnProjections.values().stream().allMatch(ProjectedColumnRepresentation::isVariable)) {
-            return Optional.empty();
+            Set<ColumnHandle> projectedColumns = ImmutableSet.copyOf(assignments.values());
+            if (hiveTableHandle.getProjectedColumns().isPresent()
+                    && hiveTableHandle.getProjectedColumns().get().containsAll(projectedColumns)) {
+                return Optional.empty();
```
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2296]: Why empty? I think this should be `hiveTableHandle.getProjectedColumns().get().equals(projectedColumns)`. Otherwise, we are pruning columns further. Please add test.

Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2296]: It should be equals, fixing
Empty because we don't need to do anything more

**Review by phd3 — COMMENTED** [review]
looks good, modulo current comments.

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java [hunk 4]
```diff
@@ -148,6 +148,10 @@ public void testPushProjection()
                     identity, "projected_variable_" + connectorNames.get(identity),
                     dereference, "projected_dereference_" + connectorNames.get(dereference),
                     constant, "projected_constant_" + connectorNames.get(constant));
+            ImmutableMap<String, ColumnHandle> expectedColumns = newNames.entrySet().stream()
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java:151]: Adde TestPruneTableScanColumns#testPushColumnPruningProjection and TestConnectorPushdownRulesWithHive#testColumnPruningProjectionPushdown

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/optimizer/TestHiveProjectionPushdownIntoTableScan.java [hunk 5]
```diff
@@ -160,7 +160,8 @@ public void testDereferencePushdown()
         assertPlan(
                 "SELECT col0.x expr_x, col0.y expr_y FROM " + testTable,
                 any(tableScan(
-                        equalTo(tableHandle.get().getConnectorHandle()),
+                        equalTo(((HiveTableHandle) tableHandle.get().getConnectorHandle())
+                                .withProjectedColumns(ImmutableSet.of(columnX, columnY))),
                         TupleDomain.all(),
                         ImmutableMap.of("col0#x", equalTo(columnX), "col0#y", equalTo(columnY)))));
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/optimizer/TestHiveProjectionPushdownIntoTableScan.java:167]: added assertions on projected columns in the following two plan assertions as well

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 6]
```diff
@@ -2463,6 +2482,7 @@ public ConnectorTableHandle makeCompatiblePartitioning(ConnectorSession session,
                 hiveTable.getAnalyzePartitionValues(),
                 hiveTable.getAnalyzeColumnNames(),
                 Optional.empty(),
+                Optional.empty(),
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: yup, added comment

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java (hunk 4 above)
Comment (losipiuk) [inline core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java:151]: declare as `Map`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 7]
```diff
@@ -2287,13 +2287,28 @@ public void validateScan(ConnectorSession session, ConnectorTableHandle tableHan
...
-            return Optional.empty();
+            Set<ColumnHandle> projectedColumns = ImmutableSet.copyOf(assignments.values());
+            if (hiveTableHandle.getProjectedColumns().isPresent()
+                    && hiveTableHandle.getProjectedColumns().get().equals(projectedColumns)) {
+                return Optional.empty();
+            }
+            return Optional.of(new ProjectionApplicationResult<>(
+                    hiveTableHandle.withProjectedColumns(projectedColumns),
+                    projections,
+                    assignments.entrySet().stream()
```
Comment (losipiuk) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2301]: Aren't you just making a strict copy of `assignments` here? Why not just use `assignments`? Or did you mean to filter out some netries?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 8]
```diff
@@ -2463,6 +2482,7 @@ public ConnectorTableHandle makeCompatiblePartitioning(ConnectorSession session,
                 hiveTable.getAnalyzePartitionValues(),
                 hiveTable.getAnalyzeColumnNames(),
                 Optional.empty(),
+                Optional.empty(), // Projected columns is used only during optimization phase of planning
```
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2485]: method doc is:
```
     * Return a table handle whose partitioning is converted to the provided partitioning handle,
     * but otherwise identical to the provided table handle.
```
so you should be able to just pass `tableHandle` projections

**Review by sopel39 — APPROVED** [review]
lgtm % @losipiuk and mine comments

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java (hunk 4 above)
Comment (raunaqmorarka) [reply core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java:151]: i was trying to be consistent with previous variable declarations here

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 7 above)
Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2301]: I am making a copy, but I need to convert `Map<String, ColumnHandle> assignments` to `List<Assignment>`.
Let me know if there is a better to do this.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 8 above)
Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2485]: I know it would be valid to pass the tableHandle projections. I just avoided it for the sake of efficiency.
It appears that we've done the same for `constraintColumns` as well.

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java (hunk 4 above)
Comment (losipiuk) [reply core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java:151]: If we are declaring variables as `ImmutableMap/List/Set` then this is an ommision. Actually we should probably have a checkstyle rule for that.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 7 above)
Comment (losipiuk) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2301]: Oh I see it know. It I do not see a much nicer way to do this. What would be more readable to me is to extract this as a named `List<Assignment> assignmentsList` variable.

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java (hunk 4 above)
Comment (raunaqmorarka) [reply core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java]: Changed mine to `Map`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 7 above)
Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: updated to extract `List<Assignment> assignmentsList` variable

**Review by sopel39 — APPROVED** [review]

**Review by losipiuk — APPROVED** [review]
